### PR TITLE
Refactor MoE

### PR DESCRIPTION
**Pull request by kpaigwar**
This pull request introduces significant refactoring to the `MoE` model configuration. It moves CCLs to create state and output tensor buffers are created at runtime based on dynamic batch/seq_len.

### Refactoring and Simplification:
- **Removed Redundant Parameters in Configuration Methods**: The `batch_size` and `seq_len` parameters were removed from `prefill_model_config` and `decode_model_config` methods, simplifying their signatures. The `ccl` parameter was also removed from these methods and added only where necessary. 
- **Restructured Model State Creation**: A new `create_state` method was added to encapsulate CCL-related configurations and replace the previous `model_config` method. It ensures that model state creation is more modular and focused. 

### Improved Runtime Buffer Management:
- **New `create_runtime_output_buffers` Method**: This method dynamically creates runtime output buffers for dispatch and combine operations, improving memory management and making the runtime logic more modular. 
- **Updated Runtime Logic to Use Buffers**: The forward pass now uses the output buffers created by `create_runtime_output_buffers`, replacing hardcoded configurations and tensor allocations. 

### Ticket
Link to Github Issue

### Problem description
Provide context for the problem.

### What's changed
Describe the approach used to solve the problem.
Summarize the changes made and its impact.

### Checklist
- [ ] [All post commit](https://github.com/tenstorrent/tt-metal/actions/runs/16667489068) CI passes
- [ ] [DeepSeek](https://github.com/tenstorrent/tt-metal/actions/runs/16667485141)

### CODE REVIEW

File: models/demos/deepseek_v3/tt/moe.py [hunk 1]
```diff
@@ -156,57 +112,142 @@ def model_config(
...
     @classmethod
     def forward(cls, x: ttnn.Tensor, cfg: RunDecodeConfig | RunPrefillConfig) -> tuple[ttnn.Tensor, ttnn.Tensor]:
         assert x.memory_config() == cfg["input_memory_config"]
+        seq_len = 1  # a2a dispatch and combine require DP=num_dispatch_devices, hence in prefill for bs=1, we interchange the seq_len with batch_size dimensions
+        batch_size_per_device = x.shape[
+            -2
+        ]  # Input is expected to be DP. In prefill, this is equivalent to seq_len_per_device
+        batch_size = batch_size_per_device * cfg["num_dispatch_devices"]  # Global batch size
 
+        output_buffers = cls.create_runtime_output_buffers(cfg["device"], batch_size, seq_len, cfg)
```
Comment (yieldthought) [inline models/demos/deepseek_v3/tt/moe.py]: Surenly we don't want to be doing `ttnn.from_torch` in the forward pass? How will that ever work with tracing? Shouldn't these buffers be allocated in `create_state` or something?

Comment (kpaigwar) [reply models/demos/deepseek_v3/tt/moe.py]: Yeah, I know. But these buffers are sized based on batch_size/seq_len and @barci2 don't want to  pass these runtime args as a part of model_config for vllm integration.

File: models/demos/deepseek_v3/tt/moe.py [hunk 2]
```diff
@@ -92,61 +86,23 @@ def model_config(
...
 
-        # Construct the config
         return {
-            "moe_gate": MoEGate.model_config(hf_config, mesh_device, mode),
-            "input_reshape": ReshapeConfig(shape=(batch_size_per_device, 1, seq_len, hf_config.hidden_size)),
-            "topk_indices_reshape": ReshapeConfig(
-                shape=(batch_size_per_device, 1, seq_len, hf_config.num_experts_per_tok)
-            ),
+            "device": mesh_device,
+            "num_devices": num_devices,
```
Comment (barci2) [inline models/demos/deepseek_v3/tt/moe.py]: this should be a model config arg

File: models/demos/deepseek_v3/tt/moe.py [hunk 3]
```diff
@@ -92,61 +86,23 @@ def model_config(
...
+            "num_experts_per_device": num_experts_per_device,
```
Comment (barci2) [inline models/demos/deepseek_v3/tt/moe.py]: this should be a model config arg

File: models/demos/deepseek_v3/tt/moe.py [hunk 4]
```diff
@@ -92,61 +86,23 @@ def model_config(
...
+            "expert_mapping_tensors": expert_mapping_tensors,
+            "hidden_size": hf_config.hidden_size,
```
Comment (barci2) [inline models/demos/deepseek_v3/tt/moe.py]: this should be a model config arg

File: models/demos/deepseek_v3/tt/moe.py [hunk 5]
```diff
@@ -92,61 +86,23 @@ def model_config(
...
+            "num_experts_per_tok": hf_config.num_experts_per_tok,
```
Comment (barci2) [inline models/demos/deepseek_v3/tt/moe.py]: this should be a model config arg

File: models/demos/deepseek_v3/tt/moe.py [hunk 6]
```diff
@@ -92,61 +86,23 @@ def model_config(
...
+            "num_dispatch_devices": tuple(mesh_device.shape)[0],
```
Comment (barci2) [inline models/demos/deepseek_v3/tt/moe.py]: this should be a model config arg

File: models/demos/deepseek_v3/tt/moe.py [hunk 7]
```diff
@@ -92,61 +86,23 @@ def model_config(
...
             "all_to_all_combine": AllToAllCombineConfig(
                 axis=0, num_links=1, memory_config=memory_config, global_semaphore=ccl.get_semaphore(0)
             ),
-            "all_to_all_combine_output_tensors": all_to_all_combine_output_tensors,
-            "combine_output_reshape": ReshapeConfig(
-                shape=(hf_config.num_experts_per_tok, 1, batch_size_per_device * seq_len, hf_config.hidden_size)
-            ),
-            "topk_weights_repeat": RepeatConfig(repeat_dims=ttnn.Shape((hf_config.hidden_size, 1, 1, 1))),
-            "mul_experts_output_with_weights": MulConfig(memory_config=memory_config),
             "final_output_reduce_scatter": ReduceScatterAsyncConfig(
```
Comment (barci2) [inline models/demos/deepseek_v3/tt/moe.py]: this should be a model config arg - only the "from_remote_multi_device_global_semaphore" and "to_remote_multi_device_global_semaphore" should be instantiated in the state; see https://github.com/tenstorrent/tt-metal/pull/26121 for an example

File: models/demos/deepseek_v3/tt/moe.py [hunk 8]
```diff
@@ -92,61 +86,23 @@ def model_config(
...
-            "expert_mapping_tensors": expert_mapping_tensors,
-            "all_to_all_dispatch_output_tensors": all_to_all_dispatch_output_tensors,
-            "all_to_all_dispatch_metadata_tensors": all_to_all_dispatch_metadata_tensors,
-            "all_to_all_output_reshape": ReshapeConfig(shape=(1, 1, batch_size * seq_len, hf_config.hidden_size)),
-            "activations_repeat": RepeatConfig(repeat_dims=ttnn.Shape((1, num_experts_per_device, 1, 1))),
-            "moe_experts": MoEExperts._create_model_config(hf_config, mesh_device, mode),
-            "experts_output_reshape": ReshapeConfig(
-                shape=(num_experts_per_device, batch_size, seq_len, hf_config.hidden_size)
-            ),
             "all_to_all_combine": AllToAllCombineConfig(
```
Comment (barci2) [inline models/demos/deepseek_v3/tt/moe.py]: same as ReduceScatterAsyncConfig

File: models/demos/deepseek_v3/tt/moe.py [hunk 9]
```diff
@@ -92,61 +86,23 @@ def model_config(
...
-                shape=(batch_size_per_device, 1, seq_len, hf_config.num_experts_per_tok)
-            ),
+            "device": mesh_device,
+            "num_devices": num_devices,
+            "num_experts_per_device": num_experts_per_device,
+            "expert_mapping_tensors": expert_mapping_tensors,
+            "hidden_size": hf_config.hidden_size,
+            "num_experts_per_tok": hf_config.num_experts_per_tok,
+            "num_dispatch_devices": tuple(mesh_device.shape)[0],
             "all_to_all_dispatch": AllToAllDispatchConfig(
```
Comment (barci2) [inline models/demos/deepseek_v3/tt/moe.py]: same as ReduceScatterAsyncConfig

File: models/demos/deepseek_v3/tt/moe.py [hunk 10]
```diff
@@ -156,57 +112,142 @@ def model_config(
...
+        """Generate decode configuration for this module.
+
+        Args:
+            hf_config: HuggingFace model configuration object
+            mesh_device: TTNN mesh device the model will be placed later on
+        Returns:
+            ModelDecodeConfig containing operator configurations for decode mode
+        """
+
+        if mode == "decode":
```
Comment (barci2) [inline models/demos/deepseek_v3/tt/moe.py]: maybe pass the memory config as the argument instead of the mode?

File: models/demos/deepseek_v3/tt/moe.py (hunk 1 above)
Comment (barci2) [reply models/demos/deepseek_v3/tt/moe.py]: yeah cause that's gonna limit our ability to run the model multi-user @yieldthought and I figured that functionality is more important than performance for functional bringup

File: models/demos/deepseek_v3/tt/moe.py [hunk 11]
```diff
@@ -50,36 +51,29 @@ def convert_weights(
         return weight_config
 
     @classmethod
-    def model_config(
+    def create_state(
         cls,
         hf_config: PretrainedConfig,
         mesh_device: ttnn.Device,
-        ccl: CCL1D,
         mode: str,
```
Comment (barci2) [inline models/demos/deepseek_v3/tt/moe.py]: state is shared between prefill and decode, so there should be no argument such as "mode"

File: models/demos/deepseek_v3/tt/moe.py [hunk 12]
```diff
@@ -156,57 +112,142 @@ def model_config(
                 memory_config=memory_config,
```
Comment (barci2) [inline models/demos/deepseek_v3/tt/moe.py:112]: this should be a part of the model config

File: models/demos/deepseek_v3/tt/moe.py (hunk 11 above)
Comment (kpaigwar) [reply models/demos/deepseek_v3/tt/moe.py]: done

File: models/demos/deepseek_v3/tt/moe.py (hunk 10 above)
Comment (kpaigwar) [reply models/demos/deepseek_v3/tt/moe.py]: def model_config is sharded by prefill and decode model configs
Usage of mode is fine here. We are not required to pass from test level. It is auto passed based on whether you creating prefil or decode config.

File: models/demos/deepseek_v3/tt/moe.py (hunk 9 above)
Comment (kpaigwar) [reply models/demos/deepseek_v3/tt/moe.py]: done

File: models/demos/deepseek_v3/tt/moe.py (hunk 8 above)
Comment (kpaigwar) [reply models/demos/deepseek_v3/tt/moe.py]: done

File: models/demos/deepseek_v3/tt/moe.py (hunk 7 above)
Comment (kpaigwar) [reply models/demos/deepseek_v3/tt/moe.py]: done, wish I knew this example before.

File: models/demos/deepseek_v3/tt/moe.py (hunk 6 above)
Comment (kpaigwar) [reply models/demos/deepseek_v3/tt/moe.py]: done

File: models/demos/deepseek_v3/tt/moe.py (hunk 5 above)
Comment (kpaigwar) [reply models/demos/deepseek_v3/tt/moe.py]: done

File: models/demos/deepseek_v3/tt/moe.py (hunk 4 above)
Comment (kpaigwar) [reply models/demos/deepseek_v3/tt/moe.py]: done

**Comment by kpaigwar** [comment]
@barci2 I addressed your comments, please review it again.

File: models/demos/deepseek_v3/tt/moe.py (hunk 1 above)
Comment (kpaigwar) [reply models/demos/deepseek_v3/tt/moe.py]: I will keep them runtime for now and let @barci2 decide, how he wants to deal with dynamic buffers later.

File: models/demos/deepseek_v3/tt/moe.py [hunk 13]
```diff
@@ -92,121 +81,175 @@ def model_config(
             memory_config=ttnn.DRAM_MEMORY_CONFIG,
             layout=ttnn.ROW_MAJOR_LAYOUT,
         )
+
+        return {
+            "expert_mapping_tensors": expert_mapping_tensors,
+            # CCL-specific parameters (semaphores and num_links)
+            "all_to_all_dispatch": {
+                "global_semaphore": ccl.get_semaphore(0),
```
Comment (avoraTT) [inline models/demos/deepseek_v3/tt/moe.py]: Is the CCL here happening on cluster axis 0?

**Review by avoraTT — APPROVED** [review]
Please run pipelins

File: models/demos/deepseek_v3/tt/moe.py (hunk 13 above)
Comment (kpaigwar) [reply models/demos/deepseek_v3/tt/moe.py]: Yeah, dispatch between rows i.e. axis=0

File: models/demos/deepseek_v3/tt/moe.py [hunk 14]
```diff
@@ -50,36 +51,24 @@ def convert_weights(
         return weight_config
```
Comment (barci2) [inline models/demos/deepseek_v3/tt/moe.py]: perhaps put the submodules in separate folders, to avoid name clashes? Just use `/ "moe_gate"` or something along those lines

File: models/demos/deepseek_v3/tt/moe.py (hunk 10 above)
Comment (barci2) [reply models/demos/deepseek_v3/tt/moe.py]: Yes, but passing the memory config directly is more general for free and models the function better. There is nothing inherent to the config that requires it to be only for prefill or decode. My point is that you use the "mode" parameter only in this if statement, so might as well eliminate the additional conversion from mode to memory_config and put that in the methods for creating the respective configs.

**Comment by barci2** [comment]
Thanks Kartik, overall very good work! Just the 2 additional comments and I think it'll be ready.

File: models/demos/deepseek_v3/tt/moe.py (hunk 10 above)
Comment (kpaigwar) [reply models/demos/deepseek_v3/tt/moe.py]: Actually, here memory_config is either L1 or DRAM interleaved based on Prefill or Decode but in future we may have unique sharding memory config for each Ops based on the mode.

Comment (kpaigwar) [reply models/demos/deepseek_v3/tt/moe.py]: Don't wanna pass memory config as a argument because it will be misleading to say one memory config for the model when each Op can have it own mem_cfg

Comment (barci2) [reply models/demos/deepseek_v3/tt/moe.py]: alright, makes sense I suppose, we will split that code then

File: models/demos/deepseek_v3/tt/moe.py (hunk 14 above)
Comment (barci2) [reply models/demos/deepseek_v3/tt/moe.py]: @kpaigwar bump

Comment (kpaigwar) [reply models/demos/deepseek_v3/tt/moe.py]: on it

File: models/demos/deepseek_v3/tt/moe.py [hunk 15]
```diff
@@ -44,42 +45,34 @@ def convert_weights(
     ) -> WeightConfig:
         weight_config = {}
 
-        weight_config["moe_gate"] = MoEGate.convert_weights(hf_config, state_dict, output_path, mesh_device, "gate.")
-        weight_config["moe_experts"] = MoEExperts.convert_weights(hf_config, state_dict, output_path, mesh_device)
+        # Create subdirectories for each submodule
+        moe_gate_path = output_path / "moe_gate"
```
Comment (barci2) [inline models/demos/deepseek_v3/tt/moe.py]: inline?

File: models/demos/deepseek_v3/tests/test_moe.py [hunk 16]
```diff
@@ -64,12 +64,12 @@ def test_forward_pass(
...
     if mode == "prefill":
-        model_config = MoE.prefill_model_config(hf_config, mesh_device, ccl, batch_size=seq_len, seq_len=1)
+        model_config = MoE.prefill_model_config(hf_config, mesh_device)
     else:
-        model_config = MoE.decode_model_config(hf_config, mesh_device, ccl, batch_size=seq_len)
+        model_config = MoE.decode_model_config(hf_config, mesh_device)
 
-    # Create a new model state
-    model_state = MoE.create_state(hf_config, mesh_device)
+    # Create a new model state with CCL
```
Comment (barci2) [inline models/demos/deepseek_v3/tests/test_moe.py]: could you please use the `assert_hidden_dim_pcc` and `get_model_config`?

**Review by barci2 — CHANGES_REQUESTED** [review]
new comments

**Review by barci2 — APPROVED** [review]
Accepting assuming the comments are addressed

**Comment by kpaigwar** [comment]
> Accepting assuming the comments are addressed

all are addressed now

**Comment by kpaigwar** [comment]
Had to revert this due to an accidental wrong commit which deleted some files. Relanding it into another PR
